plugins/gimp/file-jxl-save.cc

```diff
@@ -570,41 +538,41 @@ bool JpegXlSaveOpts::set_num_channels(int channels) {
       pixel_format.num_channels = 4;
       basic_info.num_color_channels = 3;
       basic_info.num_extra_channels = 1;
-      basic_info.alpha_bits = basic_info.bits_per_sample;
-      basic_info.alpha_exponent_bits = basic_info.exponent_bits_per_sample;
+      basic_info.alpha_bits = int(std::fmin(16, basic_info.bits_per_sample));
+      basic_info.alpha_exponent_bits = 0;
       break;
     default:
-      set_num_channels(3);
+      SetNumChannels(3);
   }  // switch
   return true;
-}  // JpegXlSaveOpts::set_num_channels
+}  // JpegXlSaveOpts::SetNumChannels
 
-bool JpegXlSaveOpts::update_babl_format() {
+bool JpegXlSaveOpts::UpdateBablFormat() {
   babl_format_str = babl_model_str + " " + babl_type_str;
   return true;
 }
 
-bool JpegXlSaveOpts::set_babl_model(std::string model) {
+bool JpegXlSaveOpts::SetBablModel(std::string model) {
   babl_model_str = model;
-  return update_babl_format();
+  return UpdateBablFormat();
 }
 
-bool JpegXlSaveOpts::set_babl_type(std::string type) {
+bool JpegXlSaveOpts::SetBablType(std::string type) {
   babl_type_str = type;
-  return update_babl_format();
+  return UpdateBablFormat();
 }
 
-bool JpegXlSaveOpts::set_pixel_type(int type) {
+bool JpegXlSaveOpts::SetPixelType(int type) {
   switch (type) {
     case JXL_TYPE_FLOAT16:
-      set_babl_type("half");
+      SetBablType("half");
       pixel_format.data_type = JXL_TYPE_FLOAT16;
       basic_info.bits_per_sample = 16;
       basic_info.exponent_bits_per_sample = 5;
```

Comment:
The encode API is still a bit evolving here (see https://github.com/libjxl/libjxl/pull/562), and we need to discuss what exactly we want to support.

I think it's important to distinguish the internal pixel format GIMP is using, the pixel format used to pass images to the encoder, and the bit depth signaled in the jxl bitstream.

I assume `set_babl_type` is used to ask GIMP to produce the image buffer in a certain format.
If so, then it should indeed match the `pixel_format` which is used to tell the encoder how to interpret the buffer.

But the `basic_info.bits_per_sample` is kind of independent of that. You can pass the image as 32-bit float, but still set `basic_info.bits_per_sample = 8`, for example. This field will be interpreted in two different ways, depending on whether XYB is used (i.e. lossy encoding) or not (i.e. lossless encoding):

- in lossless (or more generally, non-XYB) encoding, this is how many bits are actually stored losslessly. E.g. if the image is passed as float, but bits_per_sample = 8, then what actually gets stored are 8-bit ints. If `exponent_bits_per_sample > 0`, then what is effectively done is lossless encoding of float values. The bitstream stores these values internally as integers (the floats are reinterpreted as integers before encoding) and the decoder uses the `basic_info` header to interpret them correctly.
- in lossy (XYB) coding, these values are only informative; they are a suggestion of how to represent the data (e.g. to decide whether to store the result as 8-bit PNG or as 16-bit PNG. Lossless floats are not supported in this case, and `exponent_bits_per_sample` should be zero.

Reply:
I've been trying to pass data to the encoder in the same format that GIMP is using.  Should I always be sending float to the encoder?

Reply:
I just tried a couple combinations of basic_info and pixel_format.  Mismatched basic_info and pixel_format doesn't seem to work.

Reply:
The pixel_format has to match the format of the buffer. Basic_info can have any bits_per_sample, and shouldn't use exponent_bits_per_sample>0 unless lossless/keep_original_profile is done.

Reply:
The pixel format did match the buffer.  I'll play with it some more, but I would prefer to leave changes for a future pull request.

Reply:
Using float to transfer the image doesn't seem to simplify much, if anything.  Suppose the plugin were to transfer the image to the encoder using float.
* It looks like the transfer image has to be linear when encoding to XYB. But when encoding to non-XYB, the transfer image gamma has to match the original.
* The switch statements are still needed to set up basic_info.

If basic_info is just informative for XYB encoding, I don't understand why the encoder should reject uint32.

Reply:
It rejects it because iirc the spec requires it to be < 32 in the integer case. We want decoders to be able to use int32_t to represent the (suggested) pixel format. Otherwise implementations would need to have a special case for uint32 where they use an int64_t type internally for pixel values, which would be an annoying complication.